src/cgutils.cpp

```diff
@@ -560,14 +560,23 @@ static bool is_datatype_all_pointers(jl_datatype_t *dt)
     return true;
 }
 
-static bool is_tupletype_homogeneous(jl_svec_t *t)
+static bool is_tupletype_homogeneous(jl_svec_t *t, bool allow_va = false)
 {
     size_t i, l = jl_svec_len(t);
     if (l > 0) {
         jl_value_t *t0 = jl_svecref(t, 0);
-        if (!jl_is_leaf_type(t0))
+        if (!jl_is_leaf_type(t0)) {
+            if (allow_va && jl_is_vararg_type(t0) &&
+                  jl_is_leaf_type(jl_unwrap_vararg(t0)))
+                return true;
             return false;
+        }
         for(i=1; i < l; i++) {
+            if (allow_va && i == l - 1 && jl_is_vararg_type(jl_svecref(t,i))) {
+                if (!jl_types_equal(t0, jl_unwrap_vararg(jl_svecref(t,i))))
```

Comment:
this is not a well-formed type query (it can't be, since we've lost the UA bounds on `t` earlier)